File: 3.longest-substring-without-repeating-characters.cpp
/*
 * @lc app=leetcode id=3 lang=cpp
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
#include <stack>
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        queue<char> st;
        map<char, int> mp;
        int max;

        for(int i = 0; i < s.length(); i++)
        {
            //char not in s
            if(mp[s[i]] != 1)
            {
                st.push(s[i]);
                mp[s[i]] = 1;
            }
            //char in s
            else
            {
                int num = 0;
                int size = st.size();
                char c;
                do
                {
                    c = st.front();
                    mp[c] = 0;
                    st.pop();
                    num++;
                } while(c != s[i]);
                if(max <= size)
                    max = size;
                st.push(s[i]);
                mp[s[i]] = 1;
                
            }
        }
        int numlast = 0;
        while(!st.empty())
        {
            //cout << st.front() << endl;
            numlast++;
            st.pop();
        }
        if(max <= numlast)
            max = numlast;

        return max;
    }
};
// @lc code=end

File: 3.longest-substring-without-repeating-characters.java
/*
 * @lc app=leetcode id=3 lang=java
 *
 * [3] Longest Substring Without Repeating Characters
 */
import java.util.*;
// @lc code=start
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> mp = new HashMap<>();
        Queue<Character> que = new ArrayDeque<>();
        int max = 0;
        for (int i = 0; i < s.length(); i++) {
            Character c = Character.valueOf(s.charAt(i));
            while (mp.containsKey(c)) {
                Character dc = que.poll();
                mp.remove(dc);
            }
            mp.put(c, 1);
            que.add(c);
            max = Math.max(max, que.size());
        }
        return max;
    }
}
// @lc code=end

